Python/bytecodes.c

```diff
@@ -959,27 +959,31 @@ dummy_func(
             EXIT_IF(!PyDict_CheckExact(o));
         }
 
+        op(_GUARD_NOS_DICT_NOT_EXACT, (nos, unused -- nos, unused)) {
+            PyObject *o = PyStackRef_AsPyObjectBorrow(nos);
+            EXIT_IF(!PyDict_Check(o));
+        }
+
         op(_GUARD_TOS_DICT, (tos -- tos)) {
             PyObject *o = PyStackRef_AsPyObjectBorrow(tos);
             EXIT_IF(!PyDict_CheckExact(o));
         }
 
         macro(BINARY_OP_SUBSCR_DICT) =
-            _GUARD_NOS_DICT + unused/5 + _BINARY_OP_SUBSCR_DICT;
+            _GUARD_NOS_DICT_NOT_EXACT + unused/5 + _BINARY_OP_SUBSCR_DICT;
 
         op(_BINARY_OP_SUBSCR_DICT, (dict_st, sub_st -- res)) {
             PyObject *sub = PyStackRef_AsPyObjectBorrow(sub_st);
             PyObject *dict = PyStackRef_AsPyObjectBorrow(dict_st);
 
-            assert(PyDict_CheckExact(dict));
+            assert(PyDict_Check(dict));
+            DEOPT_IF(!Py_TYPE(dict)->tp_as_mapping);
```

Comment:
This (and the check below) should be in the guard, `_GUARD_NOS_DICT_NOT_EXACT`, not here in the action.

Is the additional check `PyDict_Check()` needed? I would have thought that `Py_TYPE(dict)->tp_as_mapping->mp_subscript == PyDict_Type.tp_as_mapping->mp_subscript` could only hold true for `dict` subclasses.